cudamapper/src/index_gpu.cuh

```diff
@@ -559,80 +570,73 @@ IndexGPU<SketchElementImpl>::IndexGPU(DefaultDeviceAllocator allocator,
                                       const std::uint64_t window_size,
                                       const bool hash_representations,
                                       const double filtering_parameter,
-                                      const cudaStream_t cuda_stream)
+                                      const cudaStream_t cuda_stream_generation,
+                                      const cudaStream_t cuda_stream_copy)
     : first_read_id_(descriptor.first_read())
     , kmer_size_(kmer_size)
     , window_size_(window_size)
     , number_of_reads_(0)
     , number_of_basepairs_in_longest_read_(0)
-    , allocator_(allocator)
-    , representations_d_(allocator)
-    , read_ids_d_(allocator)
-    , positions_in_reads_d_(allocator)
-    , directions_of_reads_d_(allocator)
-    , unique_representations_d_(allocator)
-    , first_occurrence_of_representations_d_(allocator)
-    , cuda_stream_(cuda_stream)
+    , representations_d_(0, allocator, cuda_stream_generation, cuda_stream_copy)
+    , read_ids_d_(0, allocator, cuda_stream_generation, cuda_stream_copy)
+    , positions_in_reads_d_(0, allocator, cuda_stream_generation, cuda_stream_copy)
+    , directions_of_reads_d_(0, allocator, cuda_stream_generation, cuda_stream_copy)
+    , unique_representations_d_(0, allocator, cuda_stream_generation, cuda_stream_copy)
+    , first_occurrence_of_representations_d_(0, allocator, cuda_stream_generation, cuda_stream_copy)
+    , is_ready_(false) // set to true at the end of generate_index()
+    , index_host_copy_source_(nullptr)
 {
     generate_index(parser,
                    first_read_id_,
                    first_read_id_ + descriptor.number_of_reads(),
                    hash_representations,
-                   filtering_parameter);
+                   filtering_parameter,
+                   allocator,
+                   cuda_stream_generation);
 
     // This is not completely necessary, but if removed one has to make sure that the next step
     // uses the same stream or that sync is done in caller
-    GW_CU_CHECK_ERR(cudaStreamSynchronize(cuda_stream_));
+    GW_CU_CHECK_ERR(cudaStreamSynchronize(cuda_stream_generation));
```

Comment:
It's a bit confusing that you have to call `wait_to_be_ready()` after the copy-from-host constructor, but not after the generation constructor. It would require some additional logic to check if the index was copied or generated, but it would make it more consistent, in my opinion.

Reply:
This is a fair point, however I'm not sure if such change is needed. When an index is generated that is done on GPU. When it is being copied from `IndexHostCopy` it can be assumem that some type of asynchronous copy is going to be taking place.
Note that one can still call `wait_to_be_ready()` on index crated by generating constructor and it will work just fine. On the other hand I agree that making the user call it every time makes it harder to forget.
I can add calls to `wait_to_be_ready()` after call to generating constructor in order to signal that that is a good practice. This will also enable to avoid sync at the end of index generation should we ever decide to do so

Reply:
In the end we decided to got for making `IndexGPU`'s generating constructor asynchronous and make it mandatory to call `wait_to_be_ready()` before using any `Index` object
done